### PR TITLE
SSOverlays & Hover Inventory Indicator

### PR DESCRIPTION
**Pull request by tigercat2000**
This ports the hover inventory indicator from /tg/, and by necessity, SSOverlays.
The hover inventory indicator basically shows where you are trying to
put an item, and then shows an item ghost in red or green, giving visual
feedback as to if that item will go in the slot or not.

SSOverlays is a much broader system for managing overlays, which should
eventually take over *all* overlays, but I have only ported the basic
system and implemented it where necessary for it to work for the hover
inventory items currently.

https://i.tigercat2000.net/2018/10/2018-10-08_22-48-59.webm

Redeeming with fix #9727 

:cl:
add: Hovering over an inventory slot now shows an indicator for if that item can go in that slot or not. Pretty!
wip: SSOverlays has been added, but not implemented!
/:cl:

### CODE REVIEW

**Comment by tigercat2000** [comment]
Requesting other @ParadiseSS13/paradise-maintainers confirm this as a minor feature

**Comment by tigercat2000** [comment]
Additionally, next on my agenda:
 - mind.dm refactor pt 2
 - https://github.com/tgstation/tgstation/pull/39830
 - NTSL2 
 - PiP

**Comment by variableundefined** [comment]
Confirmed. I'll take up reviewing if no one else volunteer.

File: code/controllers/subsystem/overlays.dm [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+/datum/controller/subsystem/overlays/Recover()
+	overlay_icon_state_caches = SSoverlays.overlay_icon_state_caches
+	overlay_icon_cache = SSoverlays.overlay_icon_cache
+	queue = SSoverlays.queue
+
+
+/datum/controller/subsystem/overlays/fire(resumed = FALSE, mc_check = TRUE)
+	var/list/queue = src.queue
+	var/static/count = 0
+	if (count)
```
Comment (tigercat2000) [inline code/controllers/subsystem/overlays.dm]: Reee if (

File: code/controllers/subsystem/overlays.dm [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		var/c = count
+		count = 0 //so if we runtime on the Cut, we don't try again.
+		queue.Cut(1,c+1)
```
Comment (tigercat2000) [inline code/controllers/subsystem/overlays.dm]: Fucking operator spacing all over this file

File: code/controllers/subsystem/overlays.dm [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		queue.Cut(1,count+1)
+		count = 0
+
+/proc/iconstate2appearance(icon, iconstate)
+	var/static/image/stringbro = new()
+	var/list/icon_states_cache = SSoverlays.overlay_icon_state_caches
+	var/list/cached_icon = icon_states_cache[icon]
+	if (cached_icon)
+		var/cached_appearance = cached_icon["[iconstate]"]
+		if (cached_appearance)
```
Comment (tigercat2000) [inline code/controllers/subsystem/overlays.dm]: christ /tg/

File: code/controllers/subsystem/overlays.dm [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+#undef QUEUE_FOR_COMPILE
+
+//TODO: Better solution for these?
+/image/proc/add_overlay(x)
+	overlays |= x
+
+/image/proc/cut_overlay(x)
+	overlays -= x
+
+/image/proc/cut_overlays(x)
```
Comment (Birdtalon) [inline code/controllers/subsystem/overlays.dm]: Parameter x unused?

**Comment by Birdtalon** [comment]
Looks good. Does this also pave the way for the hover tooltips they have at tg?

File: code/controllers/subsystem/overlays.dm (hunk 4 above)
Comment (variableundefined) [reply code/controllers/subsystem/overlays.dm]: Quite..sure they are used?

Comment (Birdtalon) [reply code/controllers/subsystem/overlays.dm]:
```
/image/proc/cut_overlays(x)
 	overlays.Cut()
```

Comment (variableundefined) [reply code/controllers/subsystem/overlays.dm]: ..Oh dear, I missed that.

**Pull request merged**